chain/client/src/client_actor.rs

```diff
@@ -229,56 +230,23 @@ impl Handler<NetworkClientMessages> for ClientActor {
-                        *part_id,
-                        parts.num_parts,
-                        hash,
-                    ) {
-                        Ok(part) => data.push(part),
-                        Err(e) => {
-                            error!(target: "sync", "Cannot build sync part (get_state_response_part): {}", e);
-                            return NetworkClientResponses::NoResponse;
-                        }
-                    }
-                }
-                if need_header {
-                    match self.client.chain.get_state_response_header(shard_id, hash) {
-                        Ok(header) => {
-                            return NetworkClientResponses::StateResponse(
-                                StateResponseInfo {
-                                    shard_id,
-                                    hash,
-                                    shard_state: ShardStateSyncResponse {
-                                        header: Some(header),
-                                        part_ids: parts.ids,
-                                        data,
-                                    },
-                                },
-                                route_back,
-                            );
-                        }
-                        Err(e) => {
-                            error!(target: "sync", "Cannot build sync header (get_state_response_header): {}", e);
-                            return NetworkClientResponses::NoResponse;
-                        }
-                    }
-                } else {
-                    return NetworkClientResponses::StateResponse(
-                        StateResponseInfo {
-                            shard_id,
-                            hash,
-                            shard_state: ShardStateSyncResponse {
-                                header: None,
-                                part_ids: parts.ids,
-                                data,
-                            },
-                        },
-                        route_back,
-                    );
-                }
+                let network_adapter = self.network_adapter.clone();
+                // TODO discuss vulnerability: a node can be flooded by StateRequests
+                actix::spawn(
+                    self.client
+                        .chain
+                        .get_state_response_by_request(shard_id, hash, need_header, parts)
+                        .then(move |result| {
+                            if let Ok(shard_state) = result {
+                                network_adapter.do_send(NetworkRequests::StateResponse {
+                                    response: StateResponseInfo { shard_id, hash, shard_state },
+                                    route_back,
+                                });
+                            }
+                            future::ready(())
+                        }),
+                );
```

Comment:
Check #1923 to spawn this task async. Not sure it is the best way to do it. cc: @bowenwang1996

Reply:
merged